File: 435-non-overlapping-intervals/ken/435.py
# Definition for an interval.
#class Interval(object):
#    def __init__(self, s=0, e=0):
#        self.start = s
#        self.end = e

class Solution(object):
    def eraseOverlapIntervals(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: int
        """
        length = len(intervals)

        if length == 0:
            return 0

        cnt = 0
        intervals.sort(key=lambda x : x.start)
        end = intervals[0].end

        for i in xrange(1, length):
            interval = intervals[i]

            if interval.start >= end:
                end = interval.end
            else:
                cnt += 1
                end = min(end, interval.end)

        return cnt

File: 435-non-overlapping-intervals/leochang/435.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
	int eraseOverlapIntervals(vector<Interval>& intervals) {
		if (0 == intervals.size()) return 0;
		sort(intervals.begin(), intervals.end(), comp);
		int count = 1;
		int end = intervals[0].end;
		for (int i = 1; i < intervals.size(); i++) {
			if (intervals[i].start >= end) {
				end = intervals[i].end;
				count++;
			}
		}
		return intervals.size() - count;
	}
	static bool comp(Interval& p1, Interval& p2) {
		return p1.end < p2.end;
	}
};
